### DISCUSSION TITLE
가비지컬렉터가 무엇이고 어떤 역할을 하나요?

### DISCUSSION THREAD

**Discussion by BaekYuri** [wanted-9th-3team/frontend-study#31] [프론트 기술 질문 공간]
가비지컬렉터가 무엇이고 어떤 역할을 하나요?

**Comment by michoball** [comment]
프로그래밍 언어들이 메모리를 관리하는 기법 중 하나로 JS 엔진도 가지고 있는 기능입니다. 
Js 프로그램 안에 동적으로 할당한 메모리에 대한 참조가 사라지게 되면 즉 값에 도달할 수 없게되면 그 값을 js 엔진이 찾아내서 메모리에서 지우는 것을 말합니다. 
예를들어 let name = max 라고 변수를 할당한 후 name = kang 으로 바꾸면 max에 대한 참조는 사라지게 되므로 js 엔진이 max라는 데이터를 메로리에서 삭제합니다.

**Comment by bona373737** [comment]
원시값, 객체, 함수 등은 모두 메모리에 저장되는데, 자바스크립트 엔진 내의 가비지 컬렉터가 더이상 사용되지 않는 값들을  감지하여 삭제,관리 해줍니다.  이때 더이상 사용되지 않는 것의 기준으로 도달 가능성(reachability) 이라는 개념이 사용되는데,  가령 변수에 객체를 할당한뒤 해당 변수에 일반 원시값을 재할당 할 경우, 최초에 저장된 객체는 여전히 메모리에 남아있지만 접근할 수 있는 주소값이 사라지기 때문에 더이상 접근 할수 없는 상태가 되는 것을 의미한다.

가비지 컬렉션은 엔진이 자동으로 수행하므로 개발자가 임의로 조작할 수 없다.

**Comment by WooGie911** [comment]
자바스크립트는 눈에 보이지 않는 곳에서 메모리 관리를 수행합니다.

원시값, 객체, 함수 등 우리가 만드는 모든 것은 메모리를 차지합니다. 

따라서 쓸모 없어지게 된 것들은 삭제,관리 해야합니다.

자바스크립트는 도달 가능성(reachability) 이라는 개념을 사용해 메모리 관리를 수행합니다.

‘도달 가능한(reachable)’ 값은 쉽게 말해 어떻게든 접근하거나 사용할 수 있는 값을 의미합니다.

자바스크립트 엔진 내에선 가비지 컬렉터(garbage collector)가 끊임없이 동작합니다. 

가비지 컬렉터는 모든 객체를 모니터링하고, 도달할 수 없는 객체는 삭제합니다.

**Reply by javascriptbiggosoo** [reply]
저도 모던 js 튜토리얼로 공부했는데 ㅎ

**Comment by BaekYuri** [comment]
가비지 컬렉터는 더이상 사용하지 않는 값들을 감지해 삭제하는 역할을 합니다.
자바와 자바스크립트, 파이썬에 가비지 컬렉터가 있는 것으로 알고 있습니다. 또한 가비지 컬렉터는 더이상 사용하지 않는 값들을 바로 바로 삭제하지 않고 약간의 텀을 두고 한번에 모아서 삭제하는 것으로 알고 있습니다.

**Comment by skyhanull** [comment]
가비지 컬렉터는 쓸모 없어진 객체가 차지하는 메모리를 자동으로 해제하는 것으로 
자바와 자바스크립트 등 에는 가비지컬렉터가 자동으로 관리를 해주고 c언어는 수동으로 구현해서 사용할 수 있습니다
장점으로는 완벽하게 메모리관리를 할 필요가 없는 것이고
단점으로는 언제 진행될지 예측이 어렵습니다
